pkg/rbd/rbd_util.go

```diff
@@ -107,64 +108,86 @@ var (
 
 // createImage creates a new ceph image with provision and volume options.
 func createImage(ctx context.Context, pOpts *rbdVolume, volSz int64, cr *util.Credentials) error {
-	var output []byte
-
-	image := pOpts.RbdImageName
 	volSzMiB := fmt.Sprintf("%dM", volSz)
+	options := librbd.NewRbdImageOptions()
 
 	logMsg := "rbd: create %s size %s format 2 (features: %s) using mon %s, pool %s "
 	if pOpts.DataPool != "" {
 		logMsg += fmt.Sprintf("data pool %s", pOpts.DataPool)
+		err := options.SetString(librbd.RbdImageOptionDataPool, pOpts.DataPool)
+		if err != nil {
+			return err
+		}
 	}
 	klog.V(4).Infof(util.Log(ctx, logMsg),
-		image, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
+		pOpts.RbdImageName, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
 
-	args := []string{"create", image, "--size", volSzMiB, "--pool", pOpts.Pool, "--id", cr.ID, "-m", pOpts.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature", pOpts.ImageFeatures}
-
-	if pOpts.DataPool != "" {
-		args = append(args, "--data-pool", pOpts.DataPool)
+	ioctx, err := pOpts.getIoctx(cr)
+	if err != nil {
+		return err
 	}
-	output, err := execCommand("rbd", args)
 
+	image, err := librbd.CreateWithOptions(ioctx, pOpts.RbdImageName, uint64(volSz), options)
 	if err != nil {
-		return errors.Wrapf(err, "failed to create rbd image, command output: %s", string(output))
+		return errors.Wrapf(err, "failed to create rbd image")
 	}
 
+	// image is not returned to the caller
+	// no need to Close(), the image is not open after Create()
+	_ = image
+
 	return nil
 }
 
-// rbdStatus checks if there is watcher on the image.
-// It returns true if there is a watcher on the image, otherwise returns false.
-func rbdStatus(ctx context.Context, pOpts *rbdVolume, cr *util.Credentials) (bool, string, error) {
-	var output string
-	var cmd []byte
+func (rv *rbdVolume) getConn(cr *util.Credentials) (*rados.Conn, error) {
+	args := []string{"--pool", rv.Pool, "-m", rv.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature=" + rv.ImageFeatures}
+	conn, err := rados.NewConn()
+	if err != nil {
+		return nil, err
+	}
+	err = conn.ParseCmdLineArgs(args)
+	if err != nil {
+		return nil, err
+	}
 
-	image := pOpts.RbdImageName
+	err = conn.Connect()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Shutdown()
 
-	klog.V(4).Infof(util.Log(ctx, "rbd: status %s using mon %s, pool %s"), image, pOpts.Monitors, pOpts.Pool)
-	args := []string{"status", image, "--pool", pOpts.Pool, "-m", pOpts.Monitors, "--id", cr.ID, "--keyfile=" + cr.KeyFile}
-	cmd, err := execCommand("rbd", args)
-	output = string(cmd)
+	return conn, nil
+}
 
-	if err, ok := err.(*exec.Error); ok {
-		if err.Err == exec.ErrNotFound {
-			klog.Errorf(util.Log(ctx, "rbd cmd not found"))
-			// fail fast if command not found
-			return false, output, err
-		}
+func (rv *rbdVolume) getIoctx(cr *util.Credentials) (*rados.IOContext, error) {
```

Comment:
Is the intention to request for an IOCtx each time it is required? Because, @Madhu-1 worked with an prototype like this and it did not yield any benefits to performance, as we establish the connection and io context per CLI/API call.

Instead I would like to see the credentials function 'NewUserCredentials` enhanced to store a connection and an io context, for use across any API that we call for the said request. The credentials are currently setup to use a defer based destroy where these can be garbage collected.

I had gone with the above [approach](https://github.com/ShyamsundarR/ceph-csi/commit/381dca46c37af18d9f2ade0549cc056bcb95b3ab) and it yielded much better results as the said operations are done once across the various calls being made by the driver for a particular RPC request.

Reply:
The librados/librbd C APIs are thread-safe so in theory you really only need to share a single connection. If you implement a connection pool in the CSI somewhere for a given set of cluster credentials, it could return a cached connection. We don't want to keep opening a new connection for each CSI request since for large clusters that would still take up the majority of CSI-time to bootstrap up the connection.

Reply:
I am planning to keep the ioctx around, and have it re-used. Basically one ioctx per StorageClass (options, credentials, ..) and have it garbage collected after a timeout.